Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -481,18 +481,65 @@ extension HTTPClient {
             let channel: Channel? = self.lock.withLock {
                 if !cancelled {
                     cancelled = true
-                    return self.channel
+                    return self.connection?.channel
                 }
                 return nil
             }
             channel?.triggerUserOutboundEvent(TaskCancelEvent(), promise: nil)
         }
 
         @discardableResult
-        func setChannel(_ channel: Channel) -> Channel {
+        func setConnection(_ connection: ConnectionPool.Connection) -> ConnectionPool.Connection {
             return self.lock.withLock {
-                self.channel = channel
-                return channel
+                self.connection = connection
+                return connection
+            }
+        }
+
+        func succeed(promise: EventLoopPromise<Response>?, with value: Response) {
+            if let connection = self.connection {
+                connection.channel.pipeline.removeHandler(name: "decompressHandler").recover { _ in }.flatMap {
+                    connection.channel.pipeline.removeHandler(name: "timeoutHandler")
+                }.recover { _ in }.flatMap { _ in
+                    connection.channel.pipeline.removeHandler(name: "taskHandler")
+                }.whenComplete { result in
+                    switch result {
+                    case .success:
+                        connection.release()
+                        promise?.succeed(value)
+                    case .failure(let error):
+                        fatalError("Couldn't remove taskHandler: \(error)")
+                    }
+                }
+            }
+        }
+
+        func fail(_ error: Error) {
+            if let connection = self.connection {
+                connection.channel.close().whenComplete { _ in
+                    connection.channel.pipeline.removeHandler(name: "taskHandler").whenComplete { result in
+                        switch result {
+                        case .success:
+                            connection.release()
+                            self.promise.fail(error)
+                        case .failure(let error):
+                            fatalError("Couldn't remove taskHandler: \(error)")
```

Comment:
this should probably be a precondition, wdyt @weissi ?

Reply:
could be a `preconditionFailure("...")` but I don't really mind.

Reply:
Maybe we can fail the promise instead? This way the error is catched early but the failure is recoverable?

Reply:
So since it's an internal error, I think its better not to throw as we discussed, will do a precondition failure then.